### PR TITLE
Add default stack size setting

### LINKED ISSUE #8835: Set the default stack maximum to 100
This is an exceedingly simple change, but it would make counting items much easier in our base 10 numerical system.

The only problems I could see would be mods hard-coded to expect only 99 of something, which may mean that the mod is poorly coded to begin with, and mods that explicitly set the stack maximum of a certain item to 99, which would probably only be a minor inconsistency that can be fixed easily.

This issue is a minor subfeature of the possibly more difficult to implement #5730, but I think that bringing this specific idea to attention is important regardless.

### PR DESCRIPTION
**Pull request by SmallJoker**
Fixes #8835 
EDIT by paramat: Attends to the feature request part of #5730 (also see https://github.com/minetest/minetest_game/issues/1724)

This PR adds a setting to let everyone freely decide about the maximal stack size.

## To do

This PR is Ready for Review.

## How to test

Advanced Settings -> modify -> enjoy.

### CODE REVIEW

**Comment by Calinou** [comment]
Is there an usecase for node and craftitem stack sizes being different? If not, we can merge them into a single setting.

**Comment by paramat** [comment]
Surely it's better to address this issue https://github.com/minetest/minetest/issues/5730#issuecomment-299988220 :
> This is possible already. The problem is, that the value gets ignored in some places.

?

**Comment by MoNTE48** [comment]
Wouldn't 96 be a better number?

**Comment by paramat** [comment]
Hmm actually, #5730 is over 2 years old and has no core dev support. If it was only a feaure request it would therefore have been closed by now. The only reason it's still open is due to the 'bug' part of it.

So sorry but this seems wrong to me currently :-1:

**Comment by paramat** [comment]
Edited first post, thsi doesn't fix the bug part of #5730

**Comment by paramat** [comment]
Removed my -1, sorry, to soon to do that.

Reading through the issues, it hasn't been explained why this is needed so much. Sorry it's not obvious to me, i've written lots of mods but don't do much item-focussed modding.
I see how it avoids the need to loop through all registered items to alter the stack size, but how is this different from or more important than any other item property?
At least when looping through items this is being done by a mod, which seems preferable to a problematic global server configuration setting.

If this feature is added, a global server configuration setting seems an easy to code but awkward way to do it. Such settings are not per-world, and need to be added and removed from .conf to apply or not apply the default stack_size.
Surely it is better that stack sizes are set by code in a mod?

See this comment by the author of the issues https://github.com/minetest/minetest_game/issues/1724#issuecomment-299650410
> Hmm, even better would be if a sub-game/mod could define the max size, I don't know if that is possible, but it would mean that different subgames/mod could have the stacksize set to what they want, which probably makes more sense then "forcing" one size via the engine.

**Comment by benrob0329** [comment]
Making all conf settings per-world is needed for multiple reasons, I suppose this is another.

**Comment by paramat** [comment]
Hmm but that's not the only problem with a global setting.

Consider a current mod that alters stack_max for all items.
Currently it has to loop through all previously registered items and override them, and some modders are complaining about having to do this.

However, if this feature is merged, then mods will no longer do the looped overriding. But then the alteration of stack_max is no longer automatically done by the mod, the mod documentation will have to inform the user of the mod to add a setting to .conf, with the associated problems of that, the inconvenience to the user, and the probability of the user not seeing the documentation or doing something wrong.

It's really weird to make the mod non-functional in terms of altering stack_max, seeing as that is exactly what it is meant to do.
So surprisingly, a looped override turns out to be far preferable.
Sorry but i have to :-1: this as it seems so wrong.

**Comment by benrob0329** [comment]
What?? Deprecating a mod because a setting exists does not make the mod preferable. The mod will continue to work, but will be unneeded. It's better for the user/server admin because this is a simpler and more robust alternative that can be adopted.

**Comment by paramat** [comment]
Yes ok, i accept that mods may not change their code due this.
As i explained, it may seem slightly simpler, but is actually a messy and problematic way to do it, and less robust.
Looping through items may be a minor inconvenience for the modder who writes the mod, but for a server owner it makes very little difference in convenience. Being done by a mod actually has advantages over using a global setting.

**Comment by benrob0329** [comment]
More messy? How is this more messy? It's a single setting for a highly requested feature that removes the need for carefully written overriding code by doing nothing more than checking for **1 setting**. Sure, a mod could be used per-world, *but so could this setting if anyone would get around to adding a per-world minetest.conf*. Besides the fact that if server owners A) Don't user per-server minetest.conf files already, B) Don't just continue using a mod, and C) Actually bother to change this it's much simpler and cleaner to do this way.

Now, we could add a `table` setting datatype and just have a generic `nodedef_defaults` setting, and that might be a saner way to go about this than individual settings.

File: src/defaultsettings.cpp [hunk 1]
```diff
@@ -323,6 +323,8 @@ void set_default_settings(Settings *settings)
 	// Server
 	settings->setDefault("disable_escape_sequences", "false");
 	settings->setDefault("strip_color_codes", "false");
+	settings->setDefault("stacksize_default_node", "99");
+	settings->setDefault("stacksize_default_item", "99");
```
Comment (HybridDog) [inline src/defaultsettings.cpp]: You don't need to add this to defaultsettings.cpp because it is only used in the lua files,
and maybe the settings have to be converted to a number.
```lua
tonumber(minetest.settings:get("stacksize_default_node")) or 99
```

**Comment by paramat** [comment]
BTW i removed my -1. I have concerns but it is not quite enough to officially disapprove.
I do think core devs should consider the 'messiness' of using a global setting to do this when it is best done from within a mod. After all it's not particularly complex to loop through items and override.

benrob0329 i have explained above. By 'messy' i don't mean complex, i mean 'problematic, bad approach'.
I know that many server owners may request this, but that doesn't make this the right thing to do, more has to be taken into consideration than that.
Per-world settings hasn't been implemented yet, so can't be used as an argument.

**Comment by SmallJoker** [comment]
Mods can still iterate over all registered items to override the stack size. To control the stack size by mods would only be possible by putting the value into `mod.conf`, and reading it before loading any mod.

**Comment by Wuzzy2** [comment]
Bad idea. There are just so many problems with this.

First of all, the stack size limit is NOT an UI feature, it is an important gameplay feature, that is, it's a key part in *game design*. It should normally not be tampered with by the player, at least not directly. Limiting the stack size is a key element in designing a good game as you can control how much the player can carry. It makes for VASTLY different gameplay if you set cobblestone (or whatever) stack size to 240, or to 60. It seems you all see the stack size limit only as an UI feature, but that's clearly not true.

Also note that every item can set their max. stack size individually. If in a game, every item sets their stack limit explicitly, then this setting will not work. So this whole idea is basically broken by design. Pedantically, you would have to add a setting for every node, or group of nodes … However, even that won't work, because engine does not know anything about special about the nodes, games can do whatever.

It would make more sense to leave the decision, whether to make it an user setting, to the game. Maybe as part of a difficulty setting or whatever. But adding this setting in the engine means you will deprive games of that decision, and just cause a horrible mess.

> Such settings are not per-world, and need to be added and removed from .conf to apply or not apply the default stack_size.

This is also crucial. Settings that you need to change every time you change games or worlds are just horrible. This setting would obviously need to be per-world, but that's just a tiny problem compared to the rest.

Please reject and close.

**Comment by Wuzzy2** [comment]
Oh, another problem is that it completely breaks the current safe modder's assumption that the default stack size is 99.

Now if this is turned into a setting, you are basically forcing game makers to explicitly set the stack size because you no longer have a reliable default …
As I said, different stack sizes lead to different gameplay, so if a game was designed under the assumption the default is 99, and then you introduce a setting, you are pulling the rug under the game dev. That's very rude.
In short, this just makes mod/game development pointlessly more complicated.

**Comment by benrob0329** [comment]
> First of all, the stack size limit is NOT an UI feature, it is an important gameplay feature, that is, it's a key part in game design. It should normally not be tampered with by the player

Considering the fact that i said

> No. I don't want to loop through every single item just to set the default stack limit in my game

I understand that it's up to the game. _But,_ server owners should be allowed customization, and games need to be able to change things simply. Now you did mention that games might be able to just directly change `minetest.nodedef_default` (which none of the core devs have mentioned as being possible) but that still leaves server owners up in the air (unless they're willing to hold onto a forked game).

If games already can easily set the default stack size, then this setting is best moved into MTG.

**Comment by SmallJoker** [comment]
> directly change minetest.nodedef_default (which none of the core devs have mentioned as being possible)

No-one mentioned it because the mod loading order is not guaranteed, so even if there are no dependencies it does not mean that modifying the table has an effect for all registered nodes. If games want to change the setting, they can use `game.conf` to override the default (if not specified in `minetest.conf`).

**Comment by benrob0329** [comment]
Aren't all game mods loaded first?

**Comment by SmallJoker** [comment]
```
debug_log_level = verbose
grep "Loading and running script" debug.txt >filtered.txt
Tidy.

Loading and running script from <DIR>/builtin/init.lua
Loading and running script from <DIR>/builtin/init.lua
Loading and running script from <DIR>/builtin/init.lua
Loading and running script from <DIR>/builtin/init.lua
Loading and running script from <DIR>/builtin/init.lua
Loading and running script from <DIR>/builtin/init.lua
Loading and running script from <DIR>/mods/gauges/init.lua
Loading and running script from <DIR>/mods/worldeditttt/worldedit/init.lua
Loading and running script from <DIR>/mods/worldeditttt/worldedit_commands/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/sfinv/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/sethome/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/screwdriver/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/player_api/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/default/init.lua
Loading and running script from <DIR>/mods/quest_chest/init.lua
Loading and running script from <DIR>/mods/lualandmg/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/xpanes/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/walls/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/vessels/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/fireflies/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/stairs/init.lua
Loading and running script from <DIR>/mods/lab/init.lua
Loading and running script from <DIR>/mods/darkage/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/give_initial_stuff/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/flowers/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/butterflies/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/fire/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/tnt/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/dungeon_loot/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/doors/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/creative/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/binoculars/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/carts/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/bucket/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/bones/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/boats/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/game_commands/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/dye/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/wool/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/farming/init.lua
Loading and running script from <DIR>/mods/unified_inventory/init.lua
Loading and running script from <DIR>/mods/farming_plus/init.lua
Loading and running script from <DIR>/mods/farbows/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/beds/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/spawn/init.lua
Loading and running script from <DIR>/games/minetest_game/mods/map/init.lua
```

**Comment by benrob0329** [comment]
Wow, our mod loading order is ~~trash~~ nonexistent.

**Comment by Wuzzy2** [comment]
So `minetest.nodedef_default` is useless for any real-world applications at the moment? That sucks. That would be a perfectly reasonable solution, if it would work.

It would be useful if Minetest would support a special mod for games that is guaranteed to be the first loaded mod in the game (after builtin, before any other game mods, before any user mods), so that various low-level init stuff, like setting `minetest.nodedef_default` can be done safely here. I suggest the name “`_init`”.

**Comment by SmallJoker** [comment]
@Wuzzy2 It would be possible to change the sorting algorithm a bit to take a `priority = ??` mod.conf setting into account, but if any other mod wants to be first, it would have to increase that value to something high to ensure it's *really* loaded first.
See also: CSS `z-index` property on websites. https://www.reddit.com/r/webdev/comments/cu1776/

**Comment by deezl** [comment]
My two cents: Some of the push-back on this feature request is just silly, IMHO. The default can still be 99. Allowing server owners, game-devs, etc to change it to something else by a simple conf setting seems a no-brainer. The issue of it changing game-play is precisely the issue. Changing the game-play should be up to the individual, and not limited for some arbitrary reason. I fully support this.

File: builtin/game/item.lua [hunk 2]
```diff
@@ -660,7 +660,7 @@ core.nodedef_default = {
 	inventory_image = "",
 	wield_image = "",
 	wield_scale = {x=1,y=1,z=1},
-	stack_max = 99,
+	stack_max = tonumber(minetest.settings:get("stacksize_default_node")) or 99,
```
Comment (sfan5) [inline builtin/game/item.lua:663]: This should be read/parsed once and reused.
(for readability, not performance)

File: builtin/game/item.lua [hunk 3]
```diff
@@ -724,7 +724,7 @@ core.craftitemdef_default = {
 	inventory_image = "",
 	wield_image = "",
 	wield_scale = {x=1,y=1,z=1},
-	stack_max = 99,
+	stack_max = tonumber(minetest.settings:get("stacksize_default_item")) or 99,
```
Comment (sfan5) [inline builtin/game/item.lua:727]: This should be read/parsed once and reused.
(for readability, not performance)

**Comment by Wuzzy2** [comment]
List of arguments that have not been addressed by anyone:
* Games were designed on the assumption that the stack size is 99, if not defined explicitly. With this PR, the stack size of their items can now be anything between 1 and 65535 (highly unpredictable), unless you set the stack size of all items explicitly. This is not a good basis to design games.
* It's not a per-world setting, so it messes with all games at once and you might have to change it multiple times between game sessions … Yes, I agree with paramat that this is really messy and ugly.
* It will lead to a messier game development because you'd have to inform the modder that they no longer can rely on a default value for `stack_size`. Note this PR does not include any warning in `lua_api.txt` about this game-breaking change.
* It seems to be (at least partially) a workaround for a limitation in Minetest, in that you cannot actually use the default node definition to set default node values (like `stack_size`) reliably

New arguments:
* As far I understand: This PR only affects the *default* stack size. This seems to imply that nodes with an *explicit* stack size won't be touched by the setting. If this is true, basically, this setting will behave very unpredictably for the *player* as well, because it might or might not have a real effect of the stack sizes, it only depends on whether the game has set an explicit value or not. In games like MCL2, in which all stack sizes are explicit, there's no effect, but in MTG (with most stack sizes being implicit), there will be an effect. For games in which *some* items have an explicit `stack_size`, and some have not, things might become especially weird.
* It seems like only the node stack size is affected, but not of craftitems. This makes no sense.

Sorry, the more I think about the PR, the stronger my rejection of it.

To be clear: I am not opposed to the overall *idea* of letting players overwrite the stack size or anything game-related. Yes, I am aware that this is a hot-button topic for players. But the way how the PR tries to solve the problem is not good, it's messy and creates more problems than it solves. It's messy both for the game author and player.

I still am 100% convinced this PR needs to be rejected as it's a design flaw IMO, and there needs to be a different, less messy approach.

Also, YES, I know, many of you dislike the default stack size of 99. But it's a default value that games have relied on, and it's also defined like that in `lua_api.txt`. We can debate endlessly on which default value is right, but there needs to be some default value nonetheless. No matter which value is picked, we should just stick with it. Ultimately, it's the responsibility of the game author to pick a stack size. And this totally includes the decision to stick with 99. The compability argument alone should have been enough reason to reject this PR.

**Comment by sfan5** [comment]
> Games were designed on the assumption that the stack size is 99, if not defined explicitly.

File: builtin/game/item.lua [hunk 4]
```diff
@@ -724,7 +726,7 @@ core.craftitemdef_default = {
 	inventory_image = "",
 	wield_image = "",
 	wield_scale = {x=1,y=1,z=1},
-	stack_max = 99,
+	stack_max = stacksize_default_item,
```
Comment (sfan5) [inline builtin/game/item.lua:729]: The fact that this is user-changeable should be stated in lua_api.txt too.

**Comment by sfan5** [comment]
> Games were designed on the assumption that the stack size is 99, if not defined explicitly.

Guess we can't ever change hardcoded behaviors to be configurable then? Someone might have designed their game in mind with it, after all.
Also in which situations do game coders need to make the assumption of a certain stack_max? Note that `ItemStack:get_stack_max()` exists to write portable code.

> It's not a per-world setting, so it messes with all games at once

Neither are any of the settings [here](https://github.com/minetest/minetest_game/blob/master/settingtypes.txt). Yet nobody cares that you can't set them per-world.

> It will lead to a messier game development because you'd have to inform the modder that they no longer can rely on a default value for stack_size

First, <i>modders</i> could not rely on a default stack size anyway because every item could have  a different one.
If you as a <i>game coder</i> wrote all your code with the assumption that the stack size is a certain value you might just have to fix it OR if you explicitly dislike user configurability add a check to prevent this.

> It seems to be (at least partially) a workaround for a limitation in Minetest, in that you cannot actually use the default node definition to set default node values (like stack_size) reliably

Well, yes. #8934 would be an alternative solution (except for the fact that `minetest.nodedef_default` is currently undocumented). But this is not an argument in itself.

> As far I understand: This PR only affects the default stack size. This seems to imply that nodes with an explicit stack size won't be touched by the setting.

Correct.

> this setting will behave very unpredictably for the player as well, because it might or might not have a real effect of the stack sizes, it only depends on whether the game has set an explicit value or not.

That's an issue with your game, not the engine.
If you want this setting to work as intended, your game should set `stacksize_default_node = 64`, `stacksize_default_item` in its minetest.conf and stop setting stack_size if it doesn't differ from the default.

> It seems like only the node stack size is affected, but not of craftitems. This makes no sense.

There is a setting for each.
Honestly I have no idea why you would want nodes and items to stack differently, but configurability (on this scale) has never hurt anyone. 

> Ultimately, it's the responsibility of the game author to pick a stack size.

Wasn't the user supposed to be king?
Either way this PR does allow that just fine. The game author can pick a default stack size and the user can still be allowed to change it (or not if the game author disallows this).

**Comment by Wuzzy2** [comment]
> Guess we can't ever change hardcoded behaviors to be configurable then?

No, but deliberately breaking backwards-compability isn't nice. Yes, sometimes it cannot be avoided and in this case you might argue that it's a minor breakage that can be fixed easily. It's a breakage nonetheless and you're not planning 6.0.0. Compability aside, when doing stuff like this, the alternative, new system should be an actual well-designed replacement. This PR is not well-designed.

> Also in which situations do game coders need to make the assumption of a certain stack_max?

I was arguing more from an overall balancing/gameplay standpoint rather than concrete code dependencies. That the game assumes from a *gameplay* standpoint that their stack sizes default to X.

> If you as a game coder wrote all your code with the assumption that the stack size is a certain value you might just have to fix it OR if you explicitly dislike user configurability add a check to prevent this.

I already explained earlier how much of a gameplay difference the stack size can make, so picking a stack size is part of designing the game. It is absurd to imply that designing a game around a given stack size is a bug that must be fixed.

It is false that I “explicitly dislike user configurability”. I already explained that I am not opposed to let the user overwrite things, I just think that the way it is implemented in this PR is not universal enough and kind of messy, as it highly depends on the concrete game to function at all.

That you are suggesting to game authors to add a “check to prevent this” doesn't make any sense. Either you want a setting to overwrite engine behavior, or you don't. You can't have it both ways, this is impossible.

As for me, I am not opposed to the idea in general, just to the SPECIFIC implementation presented in this PR.

> > It's not a per-world setting, so it messes with all games at once

> Neither are any of the settings here. Yet nobody cares that you can't set them per-world.

That's not true. https://github.com/minetest/minetest/issues/6966

> > this setting will behave very unpredictably for the player as well, because it might or might not have a real effect of the stack sizes, it only depends on whether the game has set an explicit value or not.

> That's an issue with your game, not the engine.
> If you want this setting to work as intended, your game should set stacksize_default_node = 64, stacksize_default_item = 64 in its minetest.conf and stop setting stack_size if it doesn't differ from the default.

I don't see how it makes anything better. What if a game has many different stack sizes, depending on the item type? This is a perfectly legitimate way to design a game. For example, a large stack size for common/cheap building blocks, a medium stack size for craftitems and a low stack size for furniture or stuff. 

And in fact, this is done in MCL2, where stack sizes of 1, 16 or 64 are used. Other games might have more “classes” of stack size, and the concept of a “default” stack size makes less and less sense the more “possible” stack sizes a game has. For such game, the setting becomes basically pointless, as only one stack size is actually changed, and the majority of items don't change at all.

So, the way this setting is designed is not well thought-out. Its usefulness strongly depends on the game. For games like MTG, this will impact everything, but for more complex games, the impact of the setting might be next to zero.

> > It seems like only the node stack size is affected, but not of craftitems. This makes no sense.

> There is a setting for each.

Oops, my mistake. Strike this argument then. Doesn't change anything else, however.

> Either way this PR does allow that just fine. The game author can pick a default stack size and the user can still be allowed to change it (or not if the game author disallows this).

But this is not true. For all items with explicit `stack_max`, the setting has no effect, thus defeating the entire point of this PR.

If you want to make the stack size modifyable by the player, it should be at least be consistent and truly affect ALL items, and not just a few select ones that just happen to use the default.

Also, consider the documentation of this:

> Specifies the default stack size for items and craftitems

This is not a honest description, because it leaves out important detail. Here's a more honest description of the setting:

> Specifies the default max. stack size for items and craftitems. It might or might not actually change the max. stack size of items, some items will not have their max. stack size modified at all, and it completely depends on internal technical details of the game's programming and there is no real rule as to which items will have their max. stack sizes changed and which not. It is even possible that no items are affected at all by this. Good luck.

Yes, that's how this setting sounds to me like. :P

Maybe I would be supportive of this PR if the concept would be rethought and the behavior of it would actually be consistent across games. No “the stack size might or might not change for certain items” BS.

File: builtin/game/item.lua (hunk 4 above)
Comment (rubenwardy) [reply builtin/game/item.lua:729]: Yes, please update line 6729 of lua_api to say that the value may be configured by default.

```
-- The default value of 99 may be configured by users using the setting "stacksize_default_item"
stack_max = 99,
```

File: builtin/settingtypes.txt [hunk 5]
```diff
@@ -952,6 +952,13 @@ serverlist_url (Serverlist URL) string servers.minetest.net
 #    Use this to stop players from being able to use color in their messages
 strip_color_codes (Strip color codes) bool false
 
+[*Stacksizes]
+
+#    Specifies the default stack size for nodes
+stacksize_default_node (Default node stack size) int 99
```
Comment (rubenwardy) [inline builtin/settingtypes.txt:958]: this split doesn't really make sense necessarily, and shouldn't be exposed to users. Also, note that nodes are items too.

`default_stack_max` would be a better name imo

**Comment by rubenwardy** [comment]
This is not an API breakage whatsoever, in no shape or form

> Oh, another problem is that it completely breaks the current safe modder's assumption that the default stack size is 99.

Bad assumption. Leaving the stack_max undefined still means that another mod or the game may change the default. Leaving it undefined says that you accept the default, whatever that may be

**Comment by sfan5** [comment]
> I already explained earlier how much of a gameplay difference the stack size can make, so picking a stack size is part of designing the game. [...]
> That you are suggesting to game authors to add a “check to prevent this” doesn't make any sense. Either you want a setting to overwrite engine behavior, or you don't. You can't have it both ways, this is impossible.

Why is that impossible? If you design a game that cannot work with custom stack sizes, you add this piece of code and users are informed about why their settings do not work:
```lua
if ( tonumber(minetest.settings:get("stacksize_default_node")) or 99 ) ~= 99 then
    minetest.log("error", "MineClone2 is written with a default stack size of 99 in mind, "..
        "which can't be changed using the stacksize_default engine settings.")
end
```

> That's not true. #6966

But this is exactly my point.
When `enable_fire`, `enable_lavacooling`, ... were added to minetest_game nobody said "you can't add these because they aren't world-configurable". That is left as a separate feature request for the engine.

> If you want to make the stack size modifyable by the player, it should be at least be consistent and truly affect ALL items, and not just a few select ones that just happen to use the default.

Surely if there was a setting to override the stack_max of all items there would be even more controversy about this PR. And doing so is not actually useful in practice.

> As for me, I am not opposed to the idea in general, just to the SPECIFIC implementation presented in this PR.
> Maybe I would be supportive of this PR if the concept would be rethought and the behavior of it would actually be consistent across games. No “the stack size might or might not change for certain items” BS.

Then what is your suggestion? Leave it up to games to define such a setting themselves?
I don't see how the engine can have an universal stack size setting that works with arbitrary complex games.

**Comment by rubenwardy** [comment]
I suggest waiting for Wuzzy to respond

File: builtin/settingtypes.txt [hunk 6]
```diff
@@ -1079,6 +1079,9 @@ map-dir (Map directory) path
 #    Setting it to -1 disables the feature.
 item_entity_ttl (Item entity TTL) int 900
 
+#    Specifies the default stack size
```
Comment (sfan5) [inline builtin/settingtypes.txt:1082]:
```suggestion
#    Specifies the default stack size of nodes, items and tools. Note that mods or games may explicitly set a stack for certain (or all) items.
```

**Comment by Wuzzy2** [comment]
Forum thread started:
https://forum.minetest.net/viewtopic.php?f=3&t=24546

**Comment by SmallJoker** [comment]
@Wuzzy2 Sorry, but even after this flood of text I don't understand why it's not possible to work around this by adding an error message to your game, or changing the default value in your game's minetest.conf setting.

> When enable_fire, enable_lavacooling, ... were added to minetest_game nobody said "you can't add these because they aren't world-configurable". That is left as a separate feature request for the engine.

What are your thoughts on this?

**Comment by rubenwardy** [comment]
For most games, it doesn't matter whether the default_stack_max is 99, 100, 64, or 128. For those where it does matter, they can check/override the value - as they would now.

This is a commonly requested feature that I don't see much harm in

**Comment by Wuzzy2** [comment]
> But this is exactly my point.
> When enable_fire, enable_lavacooling, ... were added to minetest_game nobody said "you can't add these because they aren't world-configurable". That is left as a separate feature request for the engine.

My point is that this setting is not very good as a global setting, its better to wait until per-world settings are availlable.


> @Wuzzy2 Sorry, but even after this flood of text I don't understand why it's not possible to work around this by adding an error message to your game and thus forcing the value of 99.

I do ***NOT*** want to force the value of 99 in the game. That's not what I said! It's also not about my own personal games, but about how this feature works out in general.

I don't see the point of this suggestion. When you permit games to overwrite the user decision anyway, what is the point of this PR, anyway? The point of this PR seems to be to give users customization,
but with your suggestion you're basically reverting this feature again. This doesn't make sense. Sorry, this workaround is just an ad-hoc solution, not a real solution.

Either let games force all their stack sizes, with no user intervention, or let users overwrite ALL stack sizes. This PR lets users only overwrite only *some* stack sizes, and it will be completely arbitrary (from the player's point of view) which stack sizes are hit.

Maybe you can explain me how I should change MCL2 …

If I do nothing, then the setting will do nothing in MCL2 (all stack sizes are currently explicitly defined). This is stupid, as it defeats the point of this setting.
If I make 64 the default in the game's `minetest.conf`, then the user setting will only affect those, with all items having a different stack size (16 or 1) being unaffected. This is a bit better, but still kinda stupid, because MCL2 will still lock down the stack size of 16 (I would have no choice). Explain me how I justify this to my players.

But MCL2 is only an example. Now consider a game which uses many more stack size classes. Lets say 250 for blocks, 1 for tools, 100 for food, 10 for throwables, 30 for explosives, 1000 for ammunition, 50 for potions, and so on. What would you suggest to the game author in this case? I don't see a good answer here. It's also not very clear which number to pick as default, there is just no sensible default here. And even if one is picked, this would just lead to the very weird situation that ONE item class is affected, and the rest isn't.

I have the impression this PR is only made with MTG in mind, with little consideration for anything else.

> Then what is your suggestion? Leave it up to games to define such a setting themselves?
> I don't see how the engine can have an universal stack size setting that works with arbitrary complex games.

Well, paramat (who isn't too happy with this either) suggested using a mod for this.

The problem is, any game that is more complex than “use the default stack size for all items” will be kind of screwed unless they force all stack size (and making the setting pointless). Are you suggesting that games that pass the complexity threshold of more than 2 stack sizes force their stack sizes for everything, thus defusing the setting?

Your suggestion of leaving the decision to the game actually might work and avoids all the mess/inconsistency (which is a good thing!). But this has the downside that most games might not actually offer such a setting for various reasons, which is probably not what you want.

Yeah, no. I'm still not really convinced this is a good thing. :-(

@paramat: Thoughts?

**Comment by sfan5** [comment]
> My point is that this setting is not very good as a global setting, its better to wait until per-world settings are available.

This could be implemented tomorrow if someone set aside a few hours.
So this isn't really an argument, the setting can be used just fine if it's added now even if the usability isn't ideal.

> When you permit games to overwrite the user decision anyway, what is the point of this PR, anyway? The point of this PR seems to be to give users customization,
but with your suggestion you're basically reverting this feature again. This doesn't make sense. Sorry, this workaround is just an ad-hoc solution, not a real solution.

The point is to allow the user to customize the default stack size <i>if</i> the game is written with a differing default stack_size in mind.

> Maybe you can explain me how I should change MCL2 …

The easiest is probably to leave it as is and add a warning telling users that they can't use `default_stack_max"`.

An alternative is to scale the different stack_max tiers with the set value (and set 64 as default max).
So by default 1, 16, 64. If you set `default_stack_max = 128`: 1, 32, 128.

> [...] What would you suggest to the game author in this case? I don't see a good answer here.

I'd suggest to not support the setting. I don't see the huge issue here to be honest.
Users wanting configurable stack sizes is a feature request that game authors will have to evaluate like any other (since engine facilities are insufficient).

**Comment by runsy** [comment]
Give the user the freedom to choose, but always within a limit: whatever a particular mod or the game creator decides. In the end the creator decides whether to give that freedom or not; he decides that he will give the player that option.

Wuzzy says that the user can be annoyed if that setting doesn't work for him. But in the .conf file (template) it may be indicated that this setting is not absolute, and he will have to conform, he will have no other choice. :-D

In short, I think it's a great idea.
